app/models/merchant.rb

```diff
@@ -33,4 +33,33 @@ def items_ready_to_ship
       .where(invoice_items: {status: "packaged"})
       .order('invoices.created_at ASC')
   end
+
+  def calculate_top_items
+    self.items
+        .select('items.*, SUM(invoice_items.quantity * invoice_items.unit_price) AS total_revenue')
+        .joins(:invoice_items)
+        .joins(invoices: :transactions) # Change 'invoice' to 'invoices'
+        .where(transactions: { result: 'success' })
+        .group('items.id')
+        .order('total_revenue DESC')
+        .limit(5)
+  end
+
+  def calculate_topitems_date(top_items)
+    top_items.map do |item|
```

Comment:
@mfaisalnazari be careful her because I see the `map` method and think that will be counted as using ruby

Reply:
what are you trying to do by iterating?

if you're trying to access a merchant's items, it  already has_many items association. Just need clarification why using `map`

Reply:
cuz there is 5 items i have to do that for all of five of them basically iterate for each and the look into the database for the query related to that item
maybe i can merge both methods together but i think every way you'd have to first pull the top item one by one and then find the top selling date for that item

Reply:
I understand what Faisal r doing but I also have the same question as Odell's. I personally think it is ok and I plan to do the same thing for best_day of top_merchant but I will use map on the view.